### PR TITLE
Turn on /analyze and fix resulting warnings

### PR DESCRIPTION
**Pull request by BillyONeal**
This is to resolve Microsoft-internal:

* VSO-1136305
* VSO-1136304
* VSO-1136300

(which all boil down to 'getting yelled at for not turning on `/analyze`')

Functional changes:

* Added annotations for consistency to definitions
* Added what I hope are 'obvious' `_Analysis_assume_`s
* When a function sometimes filled in a value and sometimes did not, and the parameter was annotated `_Out_`, I removed the annotation as the annotations to tell PREfast that the parameter is conditionally `_Out_` exceed any value they might give us.
* Filed #1008 for a 'probably a bug but I'm not sure' case in `_MP_Rem`
* Blindly suppressed additional warnings in ConcRT and Boost as there is nothing we can do about them here
* Changed `__crtInitializeCriticalSectionEx` to return the value of `InitializeCriticalSectionAndSpinCount` rather than always return `TRUE`; in Vista or later this is no functional change, and on XP it will pass through the result correctly. Most callers are still ignoring the return value (as they probably should, now) but this avoids a warning here.
* Changed `_Wcsxfrm` to use RAII to record when `bbuffer` is allocated rather than extra null checks.
* Changed `Xp_setw` to hoist constexpr comparisons into an `if constexpr` block.
* Added `_PThrow` null check before calling `RtlPcToFileHeader` in excptptr.cpp.

### CODE REVIEW

**Comment by StephanTLavavej** [comment]
I believe I can fix the builds.

**Comment by BillyONeal** [comment]
@StephanTLavavej Don't worry about it, I just haven't done that because I've been working on other PRs; this isn't urgent :)

**Comment by BillyONeal** [comment]
(In fact I'd rather this land after the atomic wait stuff since that adds new APIs that are likely to make PREfast unhappy)

**Comment by StephanTLavavej** [comment]
Building with `/analyze` is noticeably slower. Are there other possibilities, like making this off-by-default for developers, but with coverage in the CI, assuming that `/analyze` build breaks will be relatively rare?

**Comment by BillyONeal** [comment]
> Building with `/analyze` is noticeably slower. Are there other possibilities, like making this off-by-default for developers, but with coverage in the CI, assuming that `/analyze` build breaks will be relatively rare?

Sure, we can do that; I didn't bother before because I figure our test pass takes ~an hour and our build takes ~30 seconds, and turning this on makes it take ~45 seconds; but the test pass still dominates. Making it a CMake configurable option is probably a good idea.

**Comment by StephanTLavavej** [comment]
Pushed a fix for the build break, which was simple - define `__crtQueueUserWorkItem` with SAL, like the rest of the definitions. But then I noticed that `awint.hpp` had different SAL annotations from the WinSDK, some stylistic but some significant. I updated the SAL, changing deprecated `__meow` to modern `_Meow_` (note: not a mechanical in-place conversion; each occurrence was taken from the WinSDK) and fixing the differences. (I'm not sure if they were taken from an older WinSDK, or what. I used the WinSDK that's currently checked into the MSVC-internal repo.)

I ran full clean builds of x86/x64/arm/arm64 and they succeeded.

**Comment by BillyONeal** [comment]
Thanks for running that down Stephan :D

**Comment by StephanTLavavej** [comment]
@BillyONeal FYI, there's a merge conflict in `parallel_algorithms.cpp` now.

**Comment by StephanTLavavej** [comment]
I resolved the merge conflict; the code in `parallel_algorithms.cpp` was deleted, and I verified that this PR's addition of `_Analysis_assume_(_Kernel32);` didn't need to be moved anywhere else.

**Comment by BillyONeal** [comment]
Thanks STL

File: stl/src/filesystem.cpp [hunk 1]
```diff
@@ -750,7 +758,7 @@ __std_win_error __stdcall __std_fs_get_file_id(__std_fs_file_id* const _Id, cons
 }
 
 [[nodiscard]] __std_win_error __stdcall __std_fs_resize_file(
-    const wchar_t* const _Target, const uintmax_t _New_size) noexcept {
+    _In_z_ const wchar_t* const _Target, const uintmax_t _New_size) noexcept {
```
Comment (StephanTLavavej) [inline stl/src/filesystem.cpp:761]: I observe that this isn't marking `const uintmax_t _New_size` as `_In_` - but that has no effect for value parameters anyways. No change requested.

File: stl/src/filesystem.cpp [hunk 2]
```diff
@@ -901,8 +911,9 @@ struct alignas(long long) _Aligned_file_attrs {
     }
 };
 
-[[nodiscard]] __std_win_error __stdcall __std_fs_get_stats(const wchar_t* const _Path, __std_fs_stats* const _Stats,
-    __std_fs_stats_flags _Flags, const __std_fs_file_attr _Symlink_attribute_hint) noexcept {
+[[nodiscard]] _Success_(return == __std_win_error::_Success) __std_win_error
+    __stdcall __std_fs_get_stats(_In_z_ const wchar_t* const _Path, __std_fs_stats* const _Stats,
```
Comment (StephanTLavavej) [inline stl/src/filesystem.cpp:915]: I observe that this isn't marking the pointer `__std_fs_stats* const _Stats`.

**Comment by BillyONeal** [comment]
Thanks Curtis!

**Comment by cbezault** [comment]
I'm not sure how to verify that the changes I made make `/analyze` run in the CI since the output isn't any different...

**Comment by CaseyCarter** [comment]
> I'm not sure how to verify that the changes I made make `/analyze` run in the CI since the output isn't any different...

Temporarily Inject a code pattern that `/analyze` warns about and see if the CI breaks?

**Comment by StephanTLavavej** [comment]
The build break was that `/clr:pure` is `/std:c++14`, so using `if constexpr` in `xxxprec.hpp` triggered the Future Technology warning. The simplest way to silence this was to use the STL's header warning suppression machinery, which determines whether that warning needs to be suppressed (in addition to silencing other noise that we never care about). I didn't use the whole public header guard, just the MSVC warning fragment.

**Comment by StephanTLavavej** [comment]
Analyzing change... :robot: Beep! Thanks for this improvement, human! :rocket:

**Pull request merged**